File: api/src/main/java/gg/mineral/api/nbt/TagType.java
package gg.mineral.api.nbt;

import java.io.IOException;
import java.util.List;
import java.util.Map;

/**
 * The types of NBT tags that exist.
 */
public enum TagType {

    END("End", null, Void.class) {
        @Override
        public Tag<?> newObj(Object type) {
            return null;
        }
    },
    BYTE("Byte", ByteTag.class, byte.class) {
        @Override
        public Tag<?> newObj(Object type) {
            return new ByteTag((byte) type);
        }
    },
    SHORT("Short", ShortTag.class, short.class) {
        @Override
        public Tag<?> newObj(Object type) {
            return new ShortTag((short) type);
        }
    },
    INT("Int", IntTag.class, int.class) {
        @Override
        public Tag<?> newObj(Object type) {
            return new IntTag((int) type);
        }
    },
    LONG("Long", LongTag.class, long.class) {
        @Override
        public Tag<?> newObj(Object type) {
            return new LongTag((long) type);
        }
    },
    FLOAT("Float", FloatTag.class, float.class) {
        @Override
        public Tag<?> newObj(Object type) {
            return new FloatTag((float) type);
        }
    },
    DOUBLE("Double", DoubleTag.class, double.class) {
        @Override
        public Tag<?> newObj(Object type) {
            return new DoubleTag((double) type);
        }
    },
    BYTE_ARRAY("Byte_Array", ByteArrayTag.class, byte[].class) {
        @Override
        public Tag<?> newObj(Object type) {
            return new ByteArrayTag((byte[]) type);
        }
    },
    STRING("String", StringTag.class, String.class) {
        @Override
        public Tag<?> newObj(Object type) {
            return new StringTag((String) type);
        }
    },
    // javac complains about this because ListTag is generic
    @SuppressWarnings("unchecked")
    LIST("List", ListTag.class, List.class) {
        @Override
        public Tag<?> newObj(Object type) {
            return null;
        }
    },
    COMPOUND("Compound", CompoundTag.class, Map.class) {
        @Override
        public Tag<?> newObj(Object type) {
            return new CompoundTag();
        }
    },
    INT_ARRAY("Int_Array", IntArrayTag.class, int[].class) {
        @Override
        public Tag<?> newObj(Object type) {
            return new IntArrayTag((int[]) type);
        }
    };

    private final String name;
    private final Class<? extends Tag<?>> tagClass;
    private final Class<?> valueClass;

    <V, T extends Tag<? extends V>> TagType(String name, Class<T> tagClass, Class<V> valueClass) {
        // ? extends V is needed to get Compound to work for some reason
        this.name = name;
        this.tagClass = tagClass;
        this.valueClass = valueClass;
    }

    public static TagType byId(int id) {
        if (id < 0 || id >= values().length)
            return null;
        return values()[id];
    }

    static TagType byIdOrError(int id) throws IOException {
        if (id < 0 || id >= values().length)
            throw new IOException("Invalid tag type: " + id);
        return values()[id];
    }

    public byte getId() {
        return (byte) ordinal();
    }

    public String getName() {
        return name;
    }

    public Class<? extends Tag<?>> getTagClass() {
        return tagClass;
    }

    public Class<?> getValueClass() {
        return valueClass;
    }

    public abstract Tag<?> newObj(Object type);
}

File: api/src/main/kotlin/gg/mineral/api/nbt/TagType.kt
package gg.mineral.api.nbt

import java.io.IOException

/**
 * The types of NBT tags that exist.
 */
enum class TagType(val typeName: String, val tagClass: Class<out Tag<*>>?, val valueClass: Class<*>) {
    END("End", null, Void::class.java) {
        override fun newObj(type: Any): Tag<*>? {
            return null
        }
    },
    BYTE("Byte", ByteTag::class.java, Byte::class.javaPrimitiveType!!) {
        override fun newObj(type: Any): Tag<*> = ByteTag(type as Byte)
    },
    SHORT("Short", ShortTag::class.java, Short::class.javaPrimitiveType!!) {
        override fun newObj(type: Any): Tag<*> = ShortTag(type as Short)
    },
    INT("Int", IntTag::class.java, Int::class.javaPrimitiveType!!) {
        override fun newObj(type: Any): Tag<*> = IntTag(type as Int)
    },
    LONG("Long", LongTag::class.java, Long::class.javaPrimitiveType!!) {
        override fun newObj(type: Any): Tag<*> = LongTag(type as Long)
    },
    FLOAT("Float", FloatTag::class.java, Float::class.javaPrimitiveType!!) {
        override fun newObj(type: Any): Tag<*> = FloatTag(type as Float)
    },
    DOUBLE("Double", DoubleTag::class.java, Double::class.javaPrimitiveType!!) {
        override fun newObj(type: Any): Tag<*> = DoubleTag(type as Double)
    },
    BYTE_ARRAY("Byte_Array", ByteArrayTag::class.java, ByteArray::class.java) {
        override fun newObj(type: Any): Tag<*> = ByteArrayTag(*type as ByteArray)
    },
    STRING("String", StringTag::class.java, String::class.java) {
        override fun newObj(type: Any): Tag<*> = StringTag((type as String))
    },

    // javac complains about this because ListTag is generic
    LIST("List", ListTag::class.java, MutableList::class.java) {
        override fun newObj(type: Any): Tag<*>? = null
    },
    COMPOUND("Compound", CompoundTag::class.java, MutableMap::class.java) {
        override fun newObj(type: Any): Tag<*> = CompoundTag()
    },
    INT_ARRAY("Int_Array", IntArrayTag::class.java, IntArray::class.java) {
        override fun newObj(type: Any): Tag<*> = IntArrayTag(*type as IntArray)
    };

    val id: Byte
        get() = ordinal.toByte()

    abstract fun newObj(type: Any): Tag<*>?

    companion object {
        fun byId(id: Int): TagType? {
            if (id < 0 || id >= entries.toTypedArray().size) return null
            return entries[id]
        }

        @Throws(IOException::class)
        fun byIdOrError(id: Int): TagType {
            if (id < 0 || id >= entries.toTypedArray().size) throw IOException("Invalid tag type: $id")
            return entries[id]
        }
    }
}
